File: sdks/go/promptware-go/registry.go
package promptware

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"plugin"
	"strings"
)

// ToolRegistry manages tool discovery and execution
type ToolRegistry struct {
	ToolsDir    string
	SchemasDir  string
	cache       map[string]*Tool
	schemaCache map[string]map[string]interface{}
}

// Tool represents a loaded tool adapter
type Tool struct {
	Name    string
	Version string
	Handle  func(map[string]interface{}) map[string]interface{}
	Schema  map[string]interface{}
}

// NewToolRegistry creates a new tool registry
func NewToolRegistry(promptwareRoot string) *ToolRegistry {
	return &ToolRegistry{
		ToolsDir:    filepath.Join(promptwareRoot, "tools"),
		SchemasDir:  filepath.Join(promptwareRoot, "schemas", "tools"),
		cache:       make(map[string]*Tool),
		schemaCache: make(map[string]map[string]interface{}),
	}
}

// GetTool loads a tool by name
func (r *ToolRegistry) GetTool(toolName string) (*Tool, error) {
	if tool, ok := r.cache[toolName]; ok {
		return tool, nil
	}

	// Try loading as Go plugin (compiled .so file)
	// Note: Go plugins are platform-specific and have limitations
	// For production, you'd want to use a different approach

	// For now, return error - we'll use direct compilation approach
	return nil, fmt.Errorf("tool not found: %s (Go plugin loading not implemented)", toolName)
}

// ExecuteTool executes a tool with given parameters
func (r *ToolRegistry) ExecuteTool(toolName string, params map[string]interface{}) map[string]interface{} {
	tool, err := r.GetTool(toolName)
	if err != nil {
		return map[string]interface{}{
			"ok":      false,
			"version": "v1",
			"error": map[string]interface{}{
				"code":    "E_TOOL_NOT_FOUND",
				"message": err.Error(),
			},
		}
	}

	result := tool.Handle(params)
	return result
}

// loadSchema loads JSON schema for a tool
func (r *ToolRegistry) loadSchema(toolName string) (map[string]interface{}, error) {
	if schema, ok := r.schemaCache[toolName]; ok {
		return schema, nil
	}

	schemaPaths := []string{
		filepath.Join(r.SchemasDir, toolName+".v1.json"),
		filepath.Join(r.SchemasDir, strings.ReplaceAll(toolName, "_", "-")+".v1.json"),
	}

	for _, path := range schemaPaths {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}

		var schema map[string]interface{}
		if err := json.Unmarshal(data, &schema); err != nil {
			continue
		}

		r.schemaCache[toolName] = schema
		return schema, nil
	}

	return nil, fmt.Errorf("schema not found for tool: %s", toolName)
}

// Note: Go plugin approach has limitations. For production, consider:
// 1. Compile tools directly into the binary
// 2. Use RPC/gRPC for tool communication
// 3. Use WASM for portable tool execution
// 4. Generate Go code that imports all tools at compile time

File: sdks/javascript/promptware-js/registry.js
/**
 * Tool registry for dynamic tool loading and execution.
 *
 * Provides a central registry for all available tools, similar to
 * the Python implementation in tools/registry.py.
 */
import { readFile } from 'fs/promises';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { createRequire } from 'module';
import { error, ok } from './envelope.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const require = createRequire(import.meta.url);

/**
 * Tool registry for discovering and executing tools.
 */
export class ToolRegistry {
  constructor() {
    // Point to the main Promptware tools directory
    // __dirname is promptware-js/, so dirname(__dirname) is Promptware/
    const promptwareRoot = dirname(__dirname);
    this.toolsDir = join(promptwareRoot, 'tools');
    this.schemasDir = join(promptwareRoot, 'schemas', 'tools');
    this._cache = new Map();
    this._schemaCache = new Map();
  }

  /**
   * Get tool implementation and schema by name.
   *
   * @param {string} toolName - Name of tool (e.g., 'http', 'storage')
   * @returns {Promise<object|null>} Tool object with handle function and schema
   */
  async getTool(toolName) {
    if (this._cache.has(toolName)) {
      return this._cache.get(toolName);
    }

    // Try to load from tools/<tool>/adapters/adapter_node.js
    const adapterPaths = [
      join(this.toolsDir, toolName, 'adapters', 'adapter_node.js'),
      join(this.toolsDir, toolName.replace('_', '-'), 'adapters', 'adapter_node.js')
    ];

    let toolModule = null;
    for (const adapterPath of adapterPaths) {
      try {
        // Use require to load CommonJS adapters
        toolModule = require(adapterPath);
        break;
      } catch (err) {
        // Try next path (tool may not have Node.js adapter or has missing dependencies)
        continue;
      }
    }

    if (!toolModule || !toolModule.handle) {
      return null;
    }

    // Load schema
    const schema = await this._loadSchema(toolName);

    const toolImpl = {
      handle: toolModule.handle,
      schema,
      version: toolModule.VERSION || 'v1',
      name: toolName
    };

    this._cache.set(toolName, toolImpl);
    return toolImpl;
  }

  /**
   * Load JSON schema for tool.
   *
   * @param {string} toolName - Tool name
   * @returns {Promise<object|null>} JSON schema or null
   */
  async _loadSchema(toolName) {
    if (this._schemaCache.has(toolName)) {
      return this._schemaCache.get(toolName);
    }

    const schemaPaths = [
      join(this.schemasDir, `${toolName}.v1.json`),
      join(this.schemasDir, `${toolName.replace('_', '-')}.v1.json`)
    ];

    for (const schemaPath of schemaPaths) {
      try {
        const schemaData = await readFile(schemaPath, 'utf-8');
        const schema = JSON.parse(schemaData);
        this._schemaCache.set(toolName, schema);
        return schema;
      } catch {
        // Try next path
      }
    }

    return null;
  }

  /**
   * Execute a tool with given parameters.
   *
   * @param {string} toolName - Name of tool to execute
   * @param {object} params - Parameters to pass to tool
   * @returns {Promise<object>} Tool response in envelope format
   */
  async executeTool(toolName, params) {
    const tool = await this.getTool(toolName);
    if (!tool) {
      return error('E_TOOL_NOT_FOUND', `Tool not found: ${toolName}`);
    }

    try {
      const result = await tool.handle(params);

      // Tools already return envelope format
      if (result && typeof result === 'object' && 'ok' in result) {
        return result;
      }

      // Wrap non-envelope responses
      return ok({ result });

    } catch (err) {
      return error('E_TOOL_EXECUTION', `Tool execution failed: ${err.message}`);
    }
  }
}

// Global registry instance
let _registry = null;

/**
 * Get the global tool registry instance.
 *
 * @returns {ToolRegistry} Global registry
 */
export function getRegistry() {
  if (!_registry) {
    _registry = new ToolRegistry();
  }
  return _registry;
}
